### DISCUSSION TITLE
html support in data.xml, handling of new lines and empty lines

### DISCUSSION THREAD

**Discussion by b0b23** [speedata/publisher#304] [General]
suppose we have a test data.xml like follow:
```
<data>
<text><p>
<b>P1:</b><br />
Umfang:	lose Blätter<br />
Format:	DIN A4: 21 x 29,7 cm<br />
Vorlagen: belichtungsfähige Datei gestellt<br />
Druck:	4/0-farbig Skala</p>
<br />
<br />
<br />
<p>Test</p>
<p>&nbsp;</p>
<p><br /></p>
<p><b>Variante 1:</b><br />
Papier: ca. 400 g MultiArt Silk<br />
Verarbeitung: beschneiden</p>
<p><b>Variante 2:</b><br />
Papier: 130 g MultiArt Silk<br />
Verarbeitung alt.: 24 x laminieren</p>
<p><b>Verpackung: in Stülpkarton</b><br />
Versand: frei Haus 1 Stelle Neumünster</p>
</text>
</data>
``` 

the contents of that file doesn't matter, it's just a test.
so we have a "Test" paragraph here right in the middle, surrounded by "attempts" to insert new lines before and after it. In which way should new lines be handled? the simple `<br />` tags don't seem to be right... (but logical), the nbsp (non breaking space) paragraph should be taken into account as a "empty" line imo, as should the "empty" paragraph with just a br-tag in it.

Speedata Publisher skips the "empty" paragraphs in the html. It renders none of them. The only way I can get them (all !) to render is hoops like
```
<?xml version="1.0" encoding="UTF-8"?>
<Layout
  xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Pageformat width="148mm" height="210mm"/>

  <Record element="data">
    <PlaceObject>
        <Textblock>
            <Paragraph html="all">
            <ForAll select="text">
                <ForAll select="*">
                <Value select="."></Value><Br/>
                </ForAll>
            </ForAll>
            </Paragraph>
        </Textblock>
    </PlaceObject>
  </Record>

</Layout>
```

I would expect SP to properly handle the
`<p>&nbsp;</p>`
and
`<p><br /></p>`
as proper "empty lines".

- How do you see it? What do you think about it?

the motivation for the question:

The following is not a critique on SP - not at all! - it's more about forcing a specific output, ignoring parts of the input, just because the software knows better :/
The discussion on this topic, is kinda boring, I know... My current data is saved in a database in a markdown-valid format, which gets converted to html and then fed to the SP. Markdown has some shengenians like double-white-space-enter for new lines, or "\" or stuff like that. Honestly - you simply can't expect the common users to actually use it that way. They don't care, they don't wan't to "learn" your markdown syntax.... I'm just saying... I use hard-breaks in markdown, and I want the single new lines to be actually visible in the output. Again - the users don't care - they put a new line in there, they expect to see it as such...

**Comment by pgundlach** [comment]
Your layout and data looks fine. I'd expect correct output with this layout (simplified the inner loop) or with your output without the extra `<Br />` in line 14.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Layout xmlns="urn:speedata.de:2009/publisher/en" xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Pageformat width="148mm" height="210mm" />

  <Record element="data">
    <PlaceObject>
      <Textblock>
        <Paragraph html="all">
          <Value select="text"/>
        </Paragraph>
      </Textblock>
    </PlaceObject>
  </Record>
</Layout>
```

But this doesn't work. I need to fix this. 

Both `<p>&nbsp;</p>` and `<p><br /></p>` should provide empty lines (so should `<p></p>`)

**Comment by pgundlach** [comment]
@b0b23 You can try the new version 4.3.2 which is just released. Not perfect yet but it goes in the right direction.

**Reply by b0b23** [reply]
I did try the 4.3.2.
using my data.xml above and your layout.xml new "empty" lines are set on `<p><br /></p>` and just `<br />` now.
`<p>&nbsp;</p>` and `<p></p>` do not have any effect (I'm not actually sure if later should be considered an "empty" line...).

It's definitely goes in the right direction. thank you :)

**Comment by pgundlach** [comment]
You can try 4.3.3. 

    <p></p>

Should create an empty line.

BTW: if you need proper spacing, you can try something like:

      <DefineTextformat name="text" html-vertical-spacing="all" />

(see https://doc.speedata.de/publisher/en/commandreference/definetextformat)
